File: quide-specification/src/main/java/io/gitlab/arturbosch/quide/platform/VersionAware.java
package io.gitlab.arturbosch.quide.platform;

import io.gitlab.arturbosch.quide.platform.reflect.TypeToken;
import io.gitlab.arturbosch.quide.vcs.VersionProvider;
import io.gitlab.arturbosch.quide.vcs.Versionable;

import java.util.Optional;

/**
 * @author Artur Bosch
 */
public interface VersionAware extends StorageAware {

	String LAST_VERSION = "lastVersion";
	String CURRENT_VERSION = "currentVersion";
	String VERSION_PROVIDER = "versionProviderRegistered";

	default Optional<Versionable> lastVersion() {
		return get(LAST_VERSION, TypeToken.get(Versionable.class));
	}

	default Optional<Versionable> currentVersion() {
		return get(CURRENT_VERSION, TypeToken.get(Versionable.class));
	}

	default Optional<VersionProvider> versionProvider() {
		return get(VERSION_PROVIDER, TypeToken.get(VersionProvider.class));
	}


}

File: quide-api/src/main/kotlin/io.gitlab.arturbosch.quide.api/core/VersionAware.kt
package io.gitlab.arturbosch.quide.api.core

import io.gitlab.arturbosch.quide.vcs.VersionProvider
import io.gitlab.arturbosch.quide.vcs.Versionable

/**
 * @author Artur Bosch
 */
interface VersionAware : StorageAware {

	val lastContainer get(): Versionable? = get(LAST_CONTAINER)
	val currentContainer get(): Versionable? = get(CURRENT_CONTAINER)
	val lastVersion get(): Versionable? = get(LAST_VERSION)
	val currentVersion get(): Versionable? = get(CURRENT_VERSION)
	val versionProvider get(): VersionProvider? = get(VERSION_PROVIDER)
	val isEvolutionaryAnalysis get(): Boolean = versionProvider != null

	companion object {
		const val LAST_CONTAINER = "lastContainer"
		const val CURRENT_CONTAINER = "currentContainer"
		const val LAST_VERSION = "lastVersion"
		const val CURRENT_VERSION = "currentVersion"
		const val VERSION_PROVIDER = "versionProviderRegistered"
	}
}
